src/github.com/getlantern/flashlight/pubsub/pubsub_test.go

```diff
@@ -0,0 +1,32 @@
+package pubsub
+
+import (
+	"runtime"
+	"testing"
+)
+
+type str struct {
+	Val string
+}
+
+func TestSub(t *testing.T) {
+	if Sub(&str{}, func(teststr *str) {}) != nil {
+		t.Fail()
+	}
+	if Sub(&str{}, "String") == nil {
+		t.Fail()
+	}
+}
+
+func TestPublish(t *testing.T) {
+	var received string = ""
+	Sub(&str{}, func(s *str) {
+		received = s.Val
+	})
+	Pub(&str{Val: "test"})
+
+	runtime.Gosched()
+	if received != "test" {
```

Comment:
I'm not sure that runtime.Gosched counts as synchronization over `received` though.  I agree if Go just ran these statements straightforwardly, `received` should have been mutated by the time it is checked.  And since this test passes, something close enough must be happening in the current version of Go.  All I'm saying is they reserve the right to mess with execution (e.g. for optimization) in ways that break otherwise reasonable assumptions about observability of changes to variables across goroutines.

https://golang.org/ref/mem